config/v1/types_cluster_version.go

```diff
@@ -0,0 +1,153 @@
+package v1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// ClusterVersionList is a list of ClusterVersion resources.
+// +k8s:deepcopy-gen=true
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type ClusterVersionList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+
+	Items []ClusterVersion `json:"items"`
+}
+
+// ClusterVersion is the configuration for the ClusterVersionOperator. This is where
+// parameters related to automatic updates can be set.
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen=true
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type ClusterVersion struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec is the desired state of the cluster version - the operator will work
+	// to ensure that the desired version is applied to the cluster.
+	Spec ClusterVersionSpec `json:"spec"`
+	// Status contains information about the available updates and any in-progress
+	// updates.
+	Status ClusterVersionStatus `json:"status"`
+}
+
+// ClusterVersionSpec is the desired version state of the cluster. It includes
+// the version the cluster should be at, how the cluster is identified, and
+// where the cluster should look for version updates.
+// +k8s:deepcopy-gen=true
+type ClusterVersionSpec struct {
+	// ClusterID uniquely identifies this cluster. This is expected to be
+	// an RFC4122 UUID value (xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx in
+	// hexadecimal values). This is a required field.
+	ClusterID ClusterID `json:"clusterID"`
+
+	// DesiredUpdate is an optional field that indicates the desired value of
+	// the cluster version. Setting this value will trigger an upgrade (if
+	// the current version does not match the desired version). The set of
+	// recommended update values is listed as part of available updates in
+	// status, and setting values outside that range may cause the upgrade
+	// to fail.
+	//
+	// If an upgrade fails the operator will halt and report status
+	// about the failing component. Setting the desired update value back to
+	// the previous version will cause a rollback to be attempted. Not all
+	// rollbacks will succeed.
+	//
+	// +optional
+	DesiredUpdate *Update `json:"desiredUpdate"`
+
+	// Upstream may be used to specify the preferred update server. By default
+	// it will use the appropriate update server for the cluster and region.
+	//
+	// +optional
+	Upstream *URL `json:"upstream"`
```

Comment:
Why would we distinguish between "" and nil?

Reply:
No need to, but will have to refactor the CVO for it.

Reply:
IMO `nil` is valid option, saying i do not care use your default. But "" should be an error.

Reply:
> IMO `nil` is valid option, saying i do not care use your default. But "" should be an error.

Why not make `""` mean "I don't care, use the default" and eliminate the invalid case for users to stumble on?

Reply:
Actually, do we want to be able to disable upstream?  If so, we should explicitly model that I think.

Reply:
That would probably usually be

```
UpdateServer *UpdateServer
...

type UpdateServer struct {
  // defaults to the appropriate location
  URL string
}
```

or 

```
DisableUpdates bool
Upstream URL
```

We want the unusual conditions to be opted in - bools that default to true are too confusing for users.

Reply:
> bools that default to true are too confusing for users.

bools that default to true are too confusing for **everyone**

Reply:
yes.

How about I take the follow up to add disablement as a net new field, with clear semantics that fit into "auto update" as a multi level thing:

1. I want nothing (don't even talk to anyone about updates)
2. I want to get versions from redhat [default]
3. I want to get versions from my custom server
4. I want to automatically update to those versions